prover/prover.go

```diff
@@ -308,7 +315,29 @@ func (p *Prover) onBlockProposed(
 func (p *Prover) submitProofOp(ctx context.Context, proofWithHeader *proofProducer.ProofWithHeader, isValidProof bool) {
 	p.submitProofConcurrencyGuard <- struct{}{}
 	go func() {
-		defer func() { <-p.submitProofConcurrencyGuard }()
+		defer func() {
+			<-p.submitProofConcurrencyGuard
+			delete(p.currentBlocksBeingProven, proofWithHeader.Meta.Id)
+		}()
+
+		// check if block has been verified since we started generating the proof.
+		// if it has been, transaction will revert, so we avoid gas fees even though
+		// we generated a proof we won't use.
+		isVerified, err := p.isBlockVerified(new(big.Int).SetUint64(proofWithHeader.Meta.Id))
```

Comment:
emmm i believe what we want is listening `BlockVerified` or `BlockProven` (not emitted by oracle proofs) during the proof generation (which may takes ~10 mins), and then ignore / cancel that corresponding proof generation task, since we only allow `cfg.MaxConcurrentProvingJobs` proof generation tasks at a time. I think checking this here (after proof generated (~10 mins after) and before the submission) doesn't make too much sense since the later `eth_estimateGas` call before sending the actual tx will also return this error.

Reply:
ok, makes sense!

Reply:
I think right now there is no way to "cancel" the proof generation in current zkevm implementation, maybe u can add a new method `Cancel` in the proof producer interface and just put empty functions in all implementation? Then when a block is verified, we call this `Cancel` method and then ignore this block?

Reply:
ok, i added a stub for this, can you take a look? they all just do nothing but log now.